File: frontend/src/api/ownership.ts
import { request, type QueryParams } from "@/api/client";

export type TeamStatus = "active" | "inactive";
export type PersonStatus = "active" | "inactive";
export type BusinessSystemStatus = "draft" | "active" | "inactive";

export interface TeamSummary {
  id: string;
  code: string;
  name: string;
  status: TeamStatus;
}

export interface ResponsibilityTeam extends TeamSummary {
  description: string | null;
  version: number;
  person_count: number;
  business_system_count: number;
  asset_count: number;
  created_at: string;
  updated_at: string;
}

export interface PersonSummary {
  id: string;
  employee_no: string | null;
  name: string;
  email: string | null;
  status: PersonStatus;
  team: TeamSummary;
}

export interface Person extends PersonSummary {
  phone: string | null;
  user_id: string | null;
  notes: string | null;
  version: number;
  business_system_count: number;
  asset_count: number;
  created_at: string;
  updated_at: string;
}

export interface BusinessSystem {
  id: string;
  code: string;
  name: string;
  description: string | null;
  responsible_person: PersonSummary | null;
  status: BusinessSystemStatus;
  version: number;
  asset_count: number;
  created_at: string;
  updated_at: string;
}

export interface OwnershipSummary {
  team_count: number;
  person_count: number;
  business_system_count: number;
  asset_count: number;
  complete_asset_count: number;
  unassigned_asset_count: number;
  incomplete_asset_count: number;
}

export interface PagedResult<T> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
}

export interface TeamQuery extends QueryParams {
  keyword?: string;
  status?: TeamStatus;
  has_members?: boolean;
  page?: number;
  page_size?: number;
  sort_by?: "code" | "name" | "status" | "created_at" | "updated_at";
  sort_order?: "asc" | "desc";
}

export interface PersonQuery extends QueryParams {
  keyword?: string;
  team_id?: string;
  status?: PersonStatus;
  has_email?: boolean;
  has_systems?: boolean;
  page?: number;
  page_size?: number;
  sort_by?: "name" | "employee_no" | "status" | "created_at" | "updated_at";
  sort_order?: "asc" | "desc";
}

export interface BusinessSystemQuery extends QueryParams {
  keyword?: string;
  responsible_person_id?: string;
  team_id?: string;
  status?: BusinessSystemStatus;
  has_assets?: boolean;
  page?: number;
  page_size?: number;
  sort_by?: "code" | "name" | "status" | "created_at" | "updated_at";
  sort_order?: "asc" | "desc";
}

export interface TeamCreate {
  code: string;
  name: string;
  description?: string | null;
}

export interface TeamUpdate {
  expected_version: number;
  name?: string;
  description?: string | null;
}

export interface PersonCreate {
  employee_no?: string | null;
  name: string;
  email?: string | null;
  phone?: string | null;
  team_id: string;
  user_id?: string | null;
  notes?: string | null;
  status?: PersonStatus;
}

export interface PersonUpdate extends Partial<Omit<PersonCreate, "status">> {
  expected_version: number;
}

export interface BusinessSystemCreate {
  code: string;
  name: string;
  description?: string | null;
  responsible_person_id?: string | null;
  status?: "draft" | "active";
}

export interface BusinessSystemUpdate {
  expected_version: number;
  name?: string;
  description?: string | null;
  responsible_person_id?: string | null;
}

export function getResponsibilityTeams(query: TeamQuery = {}) {
  return request<PagedResult<ResponsibilityTeam>>("/api/v1/responsibility-teams", {
    query
  });
}

export function createResponsibilityTeam(payload: TeamCreate) {
  return request<ResponsibilityTeam>("/api/v1/responsibility-teams", {
    method: "POST",
    body: payload
  });
}

export function updateResponsibilityTeam(teamId: string, payload: TeamUpdate) {
  return request<ResponsibilityTeam>(`/api/v1/responsibility-teams/${teamId}`, {
    method: "PATCH",
    body: payload
  });
}

export function setResponsibilityTeamStatus(
  teamId: string,
  action: "activate" | "deactivate",
  expectedVersion: number
) {
  return request<ResponsibilityTeam>(
    `/api/v1/responsibility-teams/${teamId}/${action}`,
    { method: "POST", body: { expected_version: expectedVersion } }
  );
}

export function transferTeamMembers(teamId: string, personIds: string[]) {
  return request<ResponsibilityTeam>(
    `/api/v1/responsibility-teams/${teamId}/transfer-members`,
    { method: "POST", body: { person_ids: personIds } }
  );
}

export function getPeople(query: PersonQuery = {}) {
  return request<PagedResult<Person>>("/api/v1/people", { query });
}

export function createPerson(payload: PersonCreate) {
  return request<Person>("/api/v1/people", { method: "POST", body: payload });
}

export function updatePerson(personId: string, payload: PersonUpdate) {
  return request<Person>(`/api/v1/people/${personId}`, {
    method: "PATCH",
    body: payload
  });
}

export function activatePerson(personId: string, expectedVersion: number) {
  return request<Person>(`/api/v1/people/${personId}/activate`, {
    method: "POST",
    body: { expected_version: expectedVersion }
  });
}

export function deactivatePerson(
  personId: string,
  expectedVersion: number,
  replacementPersonId?: string
) {
  return request<Person>(`/api/v1/people/${personId}/deactivate`, {
    method: "POST",
    body: {
      expected_version: expectedVersion,
      replacement_person_id: replacementPersonId || null
    }
  });
}

export function getBusinessSystems(query: BusinessSystemQuery = {}) {
  return request<PagedResult<BusinessSystem>>("/api/v1/business-systems", { query });
}

export function createBusinessSystem(payload: BusinessSystemCreate) {
  return request<BusinessSystem>("/api/v1/business-systems", {
    method: "POST",
    body: payload
  });
}

export function updateBusinessSystem(systemId: string, payload: BusinessSystemUpdate) {
  return request<BusinessSystem>(`/api/v1/business-systems/${systemId}`, {
    method: "PATCH",
    body: payload
  });
}

export function activateBusinessSystem(systemId: string, expectedVersion: number) {
  return request<BusinessSystem>(`/api/v1/business-systems/${systemId}/activate`, {
    method: "POST",
    body: { expected_version: expectedVersion }
  });
}

export function deactivateBusinessSystem(
  systemId: string,
  payload: {
    expected_version: number;
    replacement_system_id?: string | null;
    unassign_assets?: boolean;
  }
) {
  return request<BusinessSystem>(`/api/v1/business-systems/${systemId}/deactivate`, {
    method: "POST",
    body: payload
  });
}

export function getOwnershipSummary() {
  return request<OwnershipSummary>("/api/v1/ownership/summary");
}

File: backend/app/api/routes/ownership.py
from __future__ import annotations

from typing import Literal

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.api.deps import get_db, require_superuser
from app.db.models import User
from app.schemas.ownership import (
    BusinessSystemCreate,
    BusinessSystemDeactivateRequest,
    BusinessSystemList,
    BusinessSystemOut,
    BusinessSystemUpdate,
    OwnershipSummary,
    PersonCreate,
    PersonDeactivateRequest,
    PersonList,
    PersonOut,
    PersonUpdate,
    ResponsibilityTeamCreate,
    ResponsibilityTeamList,
    ResponsibilityTeamOut,
    ResponsibilityTeamUpdate,
    TransferMembersRequest,
    VersionedActionRequest,
)
from app.services import ownership as ownership_service
from app.services.ownership import OwnershipServiceError


router = APIRouter()


@router.get("/responsibility-teams", response_model=ResponsibilityTeamList)
async def get_responsibility_teams(
    keyword: str | None = Query(default=None, max_length=255),
    status: Literal["active", "inactive"] | None = Query(default=None),
    has_members: bool | None = Query(default=None),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=200),
    sort_by: Literal["code", "name", "status", "created_at", "updated_at"] = Query(
        default="updated_at"
    ),
    sort_order: Literal["asc", "desc"] = Query(default="desc"),
    db: Session = Depends(get_db),
) -> ResponsibilityTeamList:
    return ownership_service.list_responsibility_teams(
        db,
        keyword=keyword,
        status=status,
        has_members=has_members,
        page=page,
        page_size=page_size,
        sort_by=sort_by,
        sort_order=sort_order,
    )


@router.post(
    "/responsibility-teams",
    response_model=ResponsibilityTeamOut,
    status_code=201,
)
async def post_responsibility_team(
    payload: ResponsibilityTeamCreate,
    actor: User = Depends(require_superuser),
    db: Session = Depends(get_db),
) -> ResponsibilityTeamOut:
    return _run(
        ownership_service.create_responsibility_team,
        db,
        payload,
        actor=actor,
    )


@router.get(
    "/responsibility-teams/{team_id}",
    response_model=ResponsibilityTeamOut,
)
async def get_responsibility_team(
    team_id: str,
    db: Session = Depends(get_db),
) -> ResponsibilityTeamOut:
    return _run(ownership_service.get_responsibility_team, db, team_id)


@router.patch(
    "/responsibility-teams/{team_id}",
    response_model=ResponsibilityTeamOut,
)
async def patch_responsibility_team(
    team_id: str,
    payload: ResponsibilityTeamUpdate,
    actor: User = Depends(require_superuser),
    db: Session = Depends(get_db),
) -> ResponsibilityTeamOut:
    return _run(
        ownership_service.update_responsibility_team,
        db,
        team_id,
        payload,
        actor=actor,
    )


@router.post(
    "/responsibility-teams/{team_id}/transfer-members",
    response_model=ResponsibilityTeamOut,
)
async def post_transfer_members(
    team_id: str,
    payload: TransferMembersRequest,
    actor: User = Depends(require_superuser),
    db: Session = Depends(get_db),
) -> ResponsibilityTeamOut:
    return _run(
        ownership_service.transfer_members,
        db,
        team_id,
        payload.person_ids,
        actor=actor,
    )


@router.post(
    "/responsibility-teams/{team_id}/activate",
    response_model=ResponsibilityTeamOut,
)
async def post_activate_team(
    team_id: str,
    payload: VersionedActionRequest,
    actor: User = Depends(require_superuser),
    db: Session = Depends(get_db),
) -> ResponsibilityTeamOut:
    return _run(
        ownership_service.activate_responsibility_team,
        db,
        team_id,
        expected_version=payload.expected_version,
        actor=actor,
    )


@router.post(
    "/responsibility-teams/{team_id}/deactivate",
    response_model=ResponsibilityTeamOut,
)
async def post_deactivate_team(
    team_id: str,
    payload: VersionedActionRequest,
    actor: User = Depends(require_superuser),
    db: Session = Depends(get_db),
) -> ResponsibilityTeamOut:
    return _run(
        ownership_service.deactivate_responsibility_team,
        db,
        team_id,
        expected_version=payload.expected_version,
        actor=actor,
    )


@router.get("/people", response_model=PersonList)
async def get_people(
    keyword: str | None = Query(default=None, max_length=255),
    team_id: str | None = Query(default=None, max_length=36),
    status: Literal["active", "inactive"] | None = Query(default=None),
    has_email: bool | None = Query(default=None),
    has_systems: bool | None = Query(default=None),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=200),
    sort_by: Literal["name", "employee_no", "status", "created_at", "updated_at"] = Query(
        default="updated_at"
    ),
    sort_order: Literal["asc", "desc"] = Query(default="desc"),
    db: Session = Depends(get_db),
) -> PersonList:
    return ownership_service.list_people(
        db,
        keyword=keyword,
        team_id=team_id,
        status=status,
        has_email=has_email,
        has_systems=has_systems,
        page=page,
        page_size=page_size,
        sort_by=sort_by,
        sort_order=sort_order,
    )


@router.post("/people", response_model=PersonOut, status_code=201)
async def post_person(
    payload: PersonCreate,
    actor: User = Depends(require_superuser),
    db: Session = Depends(get_db),
) -> PersonOut:
    return _run(ownership_service.create_person, db, payload, actor=actor)


@router.get("/people/{person_id}", response_model=PersonOut)
async def get_person(
    person_id: str,
    db: Session = Depends(get_db),
) -> PersonOut:
    return _run(ownership_service.get_person, db, person_id)


@router.patch("/people/{person_id}", response_model=PersonOut)
async def patch_person(
    person_id: str,
    payload: PersonUpdate,
    actor: User = Depends(require_superuser),
    db: Session = Depends(get_db),
) -> PersonOut:
    return _run(ownership_service.update_person, db, person_id, payload, actor=actor)


@router.post("/people/{person_id}/activate", response_model=PersonOut)
async def post_activate_person(
    person_id: str,
    payload: VersionedActionRequest,
    actor: User = Depends(require_superuser),
    db: Session = Depends(get_db),
) -> PersonOut:
    return _run(
        ownership_service.activate_person,
        db,
        person_id,
        expected_version=payload.expected_version,
        actor=actor,
    )


@router.post("/people/{person_id}/deactivate", response_model=PersonOut)
async def post_deactivate_person(
    person_id: str,
    payload: PersonDeactivateRequest,
    actor: User = Depends(require_superuser),
    db: Session = Depends(get_db),
) -> PersonOut:
    return _run(
        ownership_service.deactivate_person,
        db,
        person_id,
        payload,
        actor=actor,
    )


@router.get("/business-systems", response_model=BusinessSystemList)
async def get_business_systems(
    keyword: str | None = Query(default=None, max_length=255),
    responsible_person_id: str | None = Query(default=None, max_length=36),
    team_id: str | None = Query(default=None, max_length=36),
    status: Literal["draft", "active", "inactive"] | None = Query(default=None),
    has_assets: bool | None = Query(default=None),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=200),
    sort_by: Literal["code", "name", "status", "created_at", "updated_at"] = Query(
        default="updated_at"
    ),
    sort_order: Literal["asc", "desc"] = Query(default="desc"),
    db: Session = Depends(get_db),
) -> BusinessSystemList:
    return ownership_service.list_business_systems(
        db,
        keyword=keyword,
        responsible_person_id=responsible_person_id,
        team_id=team_id,
        status=status,
        has_assets=has_assets,
        page=page,
        page_size=page_size,
        sort_by=sort_by,
        sort_order=sort_order,
    )


@router.post("/business-systems", response_model=BusinessSystemOut, status_code=201)
async def post_business_system(
    payload: BusinessSystemCreate,
    actor: User = Depends(require_superuser),
    db: Session = Depends(get_db),
) -> BusinessSystemOut:
    return _run(ownership_service.create_business_system, db, payload, actor=actor)


@router.get("/business-systems/{system_id}", response_model=BusinessSystemOut)
async def get_business_system(
    system_id: str,
    db: Session = Depends(get_db),
) -> BusinessSystemOut:
    return _run(ownership_service.get_business_system, db, system_id)


@router.patch("/business-systems/{system_id}", response_model=BusinessSystemOut)
async def patch_business_system(
    system_id: str,
    payload: BusinessSystemUpdate,
    actor: User = Depends(require_superuser),
    db: Session = Depends(get_db),
) -> BusinessSystemOut:
    return _run(
        ownership_service.update_business_system,
        db,
        system_id,
        payload,
        actor=actor,
    )


@router.post("/business-systems/{system_id}/activate", response_model=BusinessSystemOut)
async def post_activate_business_system(
    system_id: str,
    payload: VersionedActionRequest,
    actor: User = Depends(require_superuser),
    db: Session = Depends(get_db),
) -> BusinessSystemOut:
    return _run(
        ownership_service.activate_business_system,
        db,
        system_id,
        expected_version=payload.expected_version,
        actor=actor,
    )


@router.post("/business-systems/{system_id}/deactivate", response_model=BusinessSystemOut)
async def post_deactivate_business_system(
    system_id: str,
    payload: BusinessSystemDeactivateRequest,
    actor: User = Depends(require_superuser),
    db: Session = Depends(get_db),
) -> BusinessSystemOut:
    return _run(
        ownership_service.deactivate_business_system,
        db,
        system_id,
        payload,
        actor=actor,
    )


@router.get("/ownership/summary", response_model=OwnershipSummary)
async def get_ownership_summary(
    db: Session = Depends(get_db),
) -> OwnershipSummary:
    return ownership_service.get_ownership_summary(db)


def _run(function, *args, **kwargs):
    try:
        return function(*args, **kwargs)
    except OwnershipServiceError as exc:
        raise HTTPException(
            status_code=exc.status_code,
            detail={
                "code": exc.code,
                "message": exc.message,
                "details": exc.details,
            },
        ) from exc
